File: server/gcsService.js
/**
 * Google Cloud Storage Service
 * Handles uploading and downloading workflow data to/from GCS
 */

const { Storage } = require('@google-cloud/storage');
const path = require('path');

class GCSService {
    constructor() {
        this.storage = null;
        this.bucket = null;
        this.bucketName = process.env.GCS_BUCKET_NAME || 'mvp_albert';
        this.initialized = false;
    }

    /**
     * Initialize GCS connection
     */
    async init() {
        if (this.initialized) return true;

        try {
            // Check for credentials
            const keyFilename = process.env.GOOGLE_APPLICATION_CREDENTIALS;
            const projectId = process.env.GCP_PROJECT_ID;

            if (!keyFilename) {
                console.warn('[GCS] GOOGLE_APPLICATION_CREDENTIALS not set - GCS disabled');
                return false;
            }

            this.storage = new Storage({
                keyFilename,
                projectId
            });

            this.bucket = this.storage.bucket(this.bucketName);

            // Test connection
            const [exists] = await this.bucket.exists();
            if (!exists) {
                console.error(`[GCS] Bucket '${this.bucketName}' does not exist`);
                return false;
            }

            this.initialized = true;
            console.log(`[GCS] Connected to bucket: ${this.bucketName}`);
            return true;

        } catch (error) {
            console.error('[GCS] Init error:', error.message);
            return false;
        }
    }

    /**
     * Check if GCS is available
     */
    isAvailable() {
        return this.initialized;
    }

    /**
     * Upload workflow data to GCS
     * @param {string} workflowId - Workflow ID
     * @param {string} nodeId - Node ID
     * @param {object} data - Data to upload (will be JSON stringified)
     * @param {string} originalFileName - Original file name for reference
     * @returns {object} { success, gcsPath, error }
     */
    async uploadWorkflowData(workflowId, nodeId, data, originalFileName = 'data') {
        if (!this.initialized) {
            await this.init();
            if (!this.initialized) {
                return { success: false, error: 'GCS not initialized' };
            }
        }

        try {
            const timestamp = Date.now();
            const safeName = originalFileName.replace(/[^a-zA-Z0-9._-]/g, '_');
            const gcsPath = `workflows/${workflowId}/${nodeId}/${timestamp}_${safeName}.json`;
            
            const file = this.bucket.file(gcsPath);
            
            const jsonData = JSON.stringify(data);
            
            await file.save(jsonData, {
                contentType: 'application/json',
                resumable: false,
                metadata: {
                    metadata: {
                        workflowId,
                        nodeId,
                        originalFileName,
                        uploadedAt: new Date().toISOString(),
                        rowCount: Array.isArray(data) ? data.length : 1
                    }
                }
            });

            console.log(`[GCS] Uploaded ${gcsPath} (${(jsonData.length / 1024).toFixed(2)} KB)`);

            return {
                success: true,
                gcsPath,
                size: jsonData.length,
                rowCount: Array.isArray(data) ? data.length : 1
            };

        } catch (error) {
            console.error('[GCS] Upload error:', error.message);
            return { success: false, error: error.message };
        }
    }

    /**
     * Download workflow data from GCS
     * @param {string} gcsPath - Path in GCS bucket
     * @returns {object} { success, data, error }
     */
    async downloadWorkflowData(gcsPath) {
        if (!this.initialized) {
            await this.init();
            if (!this.initialized) {
                return { success: false, error: 'GCS not initialized' };
            }
        }

        try {
            const file = this.bucket.file(gcsPath);
            
            const [exists] = await file.exists();
            if (!exists) {
                return { success: false, error: 'File not found in GCS' };
            }

            const [contents] = await file.download();
            const data = JSON.parse(contents.toString());

            console.log(`[GCS] Downloaded ${gcsPath}`);

            return {
                success: true,
                data,
                rowCount: Array.isArray(data) ? data.length : 1
            };

        } catch (error) {
            console.error('[GCS] Download error:', error.message);
            return { success: false, error: error.message };
        }
    }

    /**
     * Delete workflow data from GCS
     * @param {string} gcsPath - Path in GCS bucket
     * @returns {object} { success, error }
     */
    async deleteWorkflowData(gcsPath) {
        if (!this.initialized) {
            await this.init();
            if (!this.initialized) {
                return { success: false, error: 'GCS not initialized' };
            }
        }

        try {
            const file = this.bucket.file(gcsPath);
            await file.delete({ ignoreNotFound: true });
            
            console.log(`[GCS] Deleted ${gcsPath}`);
            return { success: true };

        } catch (error) {
            console.error('[GCS] Delete error:', error.message);
            return { success: false, error: error.message };
        }
    }

    /**
     * Delete all data for a workflow
     * @param {string} workflowId - Workflow ID
     * @returns {object} { success, deletedCount, error }
     */
    async deleteWorkflowFolder(workflowId) {
        if (!this.initialized) {
            await this.init();
            if (!this.initialized) {
                return { success: false, error: 'GCS not initialized' };
            }
        }

        try {
            const prefix = `workflows/${workflowId}/`;
            const [files] = await this.bucket.getFiles({ prefix });
            
            let deletedCount = 0;
            for (const file of files) {
                await file.delete();
                deletedCount++;
            }

            console.log(`[GCS] Deleted ${deletedCount} files for workflow ${workflowId}`);
            return { success: true, deletedCount };

        } catch (error) {
            console.error('[GCS] Delete folder error:', error.message);
            return { success: false, error: error.message };
        }
    }
}

// Singleton instance
const gcsService = new GCSService();

module.exports = { gcsService, GCSService };

File: server/prefect-worker/gcs_service.py
"""
Google Cloud Storage Service for the Prefect Worker
Downloads workflow data (Excel/CSV/PDF) from GCS buckets
"""
import os
import json
from typing import Optional, Dict, Any

import config


class GCSService:
    """Handles downloading workflow data from Google Cloud Storage"""

    def __init__(self):
        self.storage_client = None
        self.bucket = None
        self.bucket_name = os.getenv("GCS_BUCKET_NAME", "mvp_albert")
        self.initialized = False

    def init(self) -> bool:
        """Initialize GCS connection. Returns True if successful."""
        if self.initialized:
            return True

        try:
            from google.cloud import storage

            credentials_path = os.getenv("GOOGLE_APPLICATION_CREDENTIALS")
            project_id = os.getenv("GCP_PROJECT_ID")

            if not credentials_path:
                print("[GCS-Python] GOOGLE_APPLICATION_CREDENTIALS not set - GCS disabled")
                return False

            self.storage_client = storage.Client(project=project_id)
            self.bucket = self.storage_client.bucket(self.bucket_name)

            # Test connection
            if not self.bucket.exists():
                print(f"[GCS-Python] Bucket '{self.bucket_name}' does not exist")
                return False

            self.initialized = True
            print(f"[GCS-Python] Connected to bucket: {self.bucket_name}")
            return True

        except ImportError:
            print("[GCS-Python] google-cloud-storage not installed - GCS disabled")
            return False
        except Exception as e:
            print(f"[GCS-Python] Init error: {e}")
            return False

    def is_available(self) -> bool:
        """Check if GCS is available"""
        return self.initialized

    def download_workflow_data(self, gcs_path: str) -> Dict[str, Any]:
        """
        Download workflow data from GCS
        
        Args:
            gcs_path: Path in GCS bucket (e.g. 'workflows/{wfId}/{nodeId}/timestamp_file.json')
            
        Returns:
            dict with keys: success, data, error, row_count
        """
        if not self.initialized:
            if not self.init():
                return {"success": False, "error": "GCS not initialized"}

        try:
            blob = self.bucket.blob(gcs_path)

            if not blob.exists():
                return {"success": False, "error": f"File not found in GCS: {gcs_path}"}

            contents = blob.download_as_text()
            data = json.loads(contents)

            row_count = len(data) if isinstance(data, list) else 1
            print(f"[GCS-Python] Downloaded {gcs_path} ({len(contents)} bytes)")

            return {
                "success": True,
                "data": data,
                "row_count": row_count
            }

        except Exception as e:
            print(f"[GCS-Python] Download error: {e}")
            return {"success": False, "error": str(e)}


# Singleton instance
gcs_service = GCSService()

